### PR TITLE
Implement cuDNN CTC loss.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Add `cuDNNCTCLoss` wrapper function.
- Add test. Verified same result as PyTorch.

Notes:
- PyTorch's CTC loss function takes a log prob. tensor and performs softmax.
  We should edit softmax to support tensors of other dimensions.
  (We can easily support tensors with rank <= 4).
- PyTorch's CTC loss function performs a reduction op at the end to produce
  a scalar value. We can simply chain the reduction op at the end for
  simplicity.

### CODE REVIEW

**Comment by dan-zheng** [comment]
Reference PyTorch program (requires PyTorch v1.0rc1):
```python
import torch
import torch.nn as nn
import torch.nn.functional as F

cuda = torch.device('cuda')

input_length = 50
batch_size = 16
alphabet_size = 20

ctc_loss = nn.CTCLoss().to(cuda)
# ctc_loss = nn.CTCLoss(reduction='none').to(cuda)

log_probs = torch.ones(input_length, batch_size, alphabet_size).log_softmax(2).detach().requires_grad_()
targets = torch.ones((batch_size), dtype=torch.long)
input_lengths = torch.full((batch_size,), input_length, dtype=torch.long)
target_lengths = torch.ones((batch_size,), dtype=torch.long)
loss = ctc_loss(log_probs, targets, input_lengths, target_lengths)
loss.backward()

print(log_probs)
print(loss)

# `log_probs`:
# tensor([[[-2.9957, -2.9957, -2.9957,  ..., -2.9957, -2.9957, -2.9957],
#          [-2.9957, -2.9957, -2.9957,  ..., -2.9957, -2.9957, -2.9957],
#          [-2.9957, -2.9957, -2.9957,  ..., -2.9957, -2.9957, -2.9957],
#          ...
#          [-2.9957, -2.9957, -2.9957,  ..., -2.9957, -2.9957, -2.9957],
#          [-2.9957, -2.9957, -2.9957,  ..., -2.9957, -2.9957, -2.9957],
#          [-2.9957, -2.9957, -2.9957,  ..., -2.9957, -2.9957, -2.9957]]],
#        requires_grad=True)

# `loss` (mean-reduced, use `reduction='none'` for same result shape as cuDNN):
# tensor(142.6360, grad_fn=<MeanBackward1>)
```

**Comment by dan-zheng** [comment]
The PyTorch DeepSpeech model [uses a different CTC loss implementation](https://github.com/SeanNaren/deepspeech.pytorch/blob/master/train.py#L9) ([`warp-ctc`](https://github.com/t-vi/warp-ctc/tree/master/src)).
We should change the PyTorch model to use the new PyTorch library implementation of CTC loss for parity.
